datacube/metadata/_stacconverter.py

```
@@ -0,0 +1,318 @@
+    """
+    Convert return type as per STAC specification
+    """
+    # In STAC spec, "instruments" have [String] type
+    if key == "eo:instrument":
+        return _as_stac_instruments(value)
+    # Convert the non-default datetimes to a string
+    elif isinstance(value, datetime.datetime) and key != "datetime":
+        return datetime_to_str(value)
+    else:
+        return value
+
+
+def eo3_to_stac_properties(dataset: Dataset, title: str | None = None) -> dict:
+    """
+    Convert EO3 properties dictionary to the Stac equivalent.
+    """
+    # Put the title at the top for document readability.
+    properties = {"title": title} if title else {}
+
+    properties.update(
+        {
+            EO3_TO_STAC_RENAMES.get(key, key): _convert_value_to_stac_type(key, val)
+            for key, val in dataset.properties.items()
+        },
+    )
+
+    return properties
+
+
+def ds2stac(
+    dataset: Dataset,
+    stac_url: str | None = None,
+    self_url: str | None = None,
+    collection_url: str | None = None,
+) -> Item:
+    if dataset.extent is not None:
+        wgs84_geometry = dataset.extent.to_crs("EPSG:4326", math.inf)
+
+        geometry = wgs84_geometry.json
+        bbox = wgs84_geometry.boundingbox.bbox
+    else:
+        geometry = None
+        bbox = None
+
+    properties = eo3_to_stac_properties(dataset, title=dataset.metadata.label)
+    properties.update(_lineage_fields(dataset))
+
+    dt = properties.get("datetime")
+
+    item = Item(
+        id=str(dataset.id),
+        datetime=dt,
+        properties=properties,
+        geometry=geometry,
+        bbox=bbox,
+        collection=dataset.product.name,
+    )
+
+    # Add links
+    for link in _stac_links(dataset, stac_url, self_url, collection_url):
+        item.links.append(link)
+
+    EOExtension.ext(item, add_if_missing=True)
+
+    if dataset.extent:
+        proj = ProjectionExtension.ext(item, add_if_missing=True)
+        if dataset.crs is None:
```

The `dataset` is a parameter for this function, why isn't this check in the else-block starting on line 234?

---

I'm not entirely sure I understand what you mean.
Adding the ProjectionExtension is optional and requires a pystac.Item to extend, so I don't see how it can be moved up.

---

I think I mixed things up and answered this in my response to the previous line, but the general pattern I reacted to was:
```python
x = thing()
if condition_not_involving_x:
   raise Error(..) (or return None/something else not involving x)
y = use(x)
```
If the code is going to throw an exception/return None and discard the work done on computing x, it's better to throw the exception/return None right away before computing x. That saves a bit of (useless) work, and the next person looking at the code doesn't have to wonder what scary side-effects that are happening in `thing()` since that obviously had to be run before the if-statement that didn't use `x`.